File: app/Models/Tenant/Complaint.php
<?php

namespace App\Models\Tenant;

use Illuminate\Database\Eloquent\Builder;
use Illuminate\Database\Eloquent\Model;
use Illuminate\Database\Eloquent\Relations\BelongsTo;
use Illuminate\Database\Eloquent\SoftDeletes;

class Complaint extends Model
{
    use SoftDeletes;

    protected $connection = 'society';

    public const STATUSES = ['open', 'in_review', 'resolved', 'closed', 'rejected'];
    public const PRIORITIES = ['low', 'medium', 'high', 'urgent'];
    public const CATEGORIES = [
        'noise', 'parking', 'security', 'staff_behavior',
        'cleanliness', 'rule_violation', 'other',
    ];

    protected $fillable = [
        'flat_id',
        'category',
        'subject',
        'description',
        'against',
        'priority',
        'status',
        'raised_by_name',
        'raised_by_phone',
        'raised_by_user_id',
        'assigned_to_user_id',
        'resolution_notes',
        'resolved_at',
        'closed_at',
        'history',
    ];

    protected $casts = [
        'resolved_at' => 'datetime',
        'closed_at' => 'datetime',
        'history' => 'array',
    ];

    /**
     * Relationships
     */
    public function flat(): BelongsTo
    {
        return $this->belongsTo(Flat::class);
    }

    public function raisedBy(): BelongsTo
    {
        return $this->belongsTo(User::class, 'raised_by_user_id');
    }

    public function assignedTo(): BelongsTo
    {
        return $this->belongsTo(User::class, 'assigned_to_user_id');
    }

    /**
     * Scopes
     */
    public function scopeOpen(Builder $query): Builder
    {
        return $query->whereIn('status', ['open', 'in_review']);
    }

    public function scopeStatus(Builder $query, ?string $status): Builder
    {
        return $status ? $query->where('status', $status) : $query;
    }

    public function scopePriority(Builder $query, ?string $priority): Builder
    {
        return $priority ? $query->where('priority', $priority) : $query;
    }

    /**
     * Append an entry to the complaint's audit trail and persist it — same
     * JSON-column trade-off as MaintenanceRequest::logHistory().
     */
    public function logHistory(string $action, ?string $note, ?int $byUserId): void
    {
        $history = $this->history ?? [];

        $history[] = [
            'at' => now()->toIso8601String(),
            'action' => $action,
            'note' => $note,
            'by_user_id' => $byUserId,
        ];

        $this->update(['history' => $history]);
    }

    public function isOpen(): bool
    {
        return in_array($this->status, ['open', 'in_review'], true);
    }
}

File: mobile/lib/features/complaints/data/complaint.dart
import '../../../core/models/flat.dart';

class Complaint {
  Complaint({
    required this.id,
    required this.category,
    required this.subject,
    required this.description,
    this.against,
    required this.priority,
    required this.status,
    this.flat,
    this.resolutionNotes,
    this.resolvedAt,
    this.closedAt,
    this.createdAt,
  });

  factory Complaint.fromJson(Map<String, dynamic> json) {
    return Complaint(
      id: json['id'] as int,
      category: json['category'] as String,
      subject: json['subject'] as String,
      description: json['description'] as String? ?? '',
      against: json['against'] as String?,
      priority: json['priority'] as String,
      status: json['status'] as String,
      flat: json['flat'] != null ? Flat.fromJson(json['flat'] as Map<String, dynamic>) : null,
      resolutionNotes: json['resolution_notes'] as String?,
      resolvedAt: json['resolved_at'] as String?,
      closedAt: json['closed_at'] as String?,
      createdAt: json['created_at'] as String?,
    );
  }

  final int id;
  final String category;
  final String subject;
  final String description;
  final String? against;
  final String priority;
  final String status;
  final Flat? flat;
  final String? resolutionNotes;
  final String? resolvedAt;
  final String? closedAt;
  final String? createdAt;

  static const categories = [
    'noise', 'parking', 'security', 'staff_behavior',
    'cleanliness', 'rule_violation', 'other',
  ];

  static const priorities = ['low', 'medium', 'high', 'urgent'];
}
